enhancements/microshift/low-latency-workloads-on-microshift.md

```diff
@@ -0,0 +1,612 @@
+---
+title: low-latency-workloads-on-microshift
+authors:
+  - "@pmtk"
+reviewers:
+  - "@sjug, Performance and Scalability expert"
+  - "@DanielFroehlich, PM"
+  - "@jogeo, QE lead"
+  - "@eslutsky, working on MicroShift workload partitioning"
+  - "@pacevedom, MicroShift lead"
+approvers:
+  - "@jerpeter1"
+api-approvers:
+  - TBD
+creation-date: 2024-06-12
+last-updated: 2024-06-12
+tracking-link:
+  - USHIFT-2981
+---
+
+# Low Latency workloads on MicroShift
+
+## Summary
+
+This enhancement describes how low latency workloads will be supported on MicroShift hosts.
+
+
+## Motivation
+
+Some customers want to run latency sensitive workload like software defined PLCs.
+
+Currently it's possible, but requires substantial amount of knowledge to correctly configure
+all involved components whereas in OpenShift everything is abstracted using Node Tuning Operator's
+PerformanceProfile CR.
+Therefore, this enhancement focuses on making this configuration easy for customers by shipping
+ready to use packages to kickstart customers' usage of low latency workloads.
+
+
+### User Stories
+
+* As a MicroShift administrator, I want to configure MicroShift host,
+  so that I can run low latency workloads.
+
+
+### Goals
+
+Provide relatively easy way to configure system for low latency workload running on MicroShift:
+- Prepare low latency TuneD profile for MicroShift
+- Prepare necessary CRI-O configurations
+- Allow configuration of Kubelet via MicroShift config
+- Add small systemd daemon to enable TuneD profile and (optionally) reboot the host if the kernel
+  arguments change to make them effective
+  
+
+### Non-Goals
+
+- Workload partitioning (i.e. pinning MicroShift control plane components)
+- Duplicate all capabilities of Node Tuning Operator
+
+
+## Proposal
+
+To ease configuration of the system for running low latency workloads on MicroShift following
+parts need to be put in place:
+- TuneD profile
+- CRI-O configuration + Kubernetes' RuntimeClass
+- Kubelet configuration (CPU, Memory, and Topology Managers)
+- Small systemd daemon to activate TuneD profile on boot and reboot the host if the kernel args
```

Comment:
```suggestion
- A systemd daemon to activate TuneD profile on boot and reboot the host if the kernel args
```

Reply:
What does actually trigger the reboot? 
We want users to stay in control of the reboot. E.g. change of a config file should not lead to immediate unwanted reboot (and maybe not comming back if kernel args prohibit start)

Reply:
THis is explained below:
```
1. `microshift-tuned.service` starts (after `tuned.service`, before `microshift.service`):
   - Saves current kernel args
   - Applies tuned `microshift-low-latency` profile
   - Verifies expected kernel args
     - ostree: `rpm-ostree kargs`
     - rpm: `grubby`
   - If the current and expected kernel args are different, reboot the node
```

Reply:
If this is too much, then should users reboot the host themselves? Then we can drop the idea of this `microshift-tuned.service`

Reply:
Disregard my comment, I was not thinking it through. 
Actually, for a fleet of edge devices, we want and need this auto-apply and reboot. Either customers add the kernel args to their blueprint, and no further action is happening. 
Or they did forget that, then we have to apply those using the mechanisms stated below (e.g. `rpm-ostree kargs`).
Just as a safety net to disable a auto-reboot (e.g. with a config option like `reboot_after_apply = false`) should be there.